File: smells/insufficient/quixotic/quixotic.rb
# Subject under test
def rank_hotel_review(user, title, text, stars)
  rank = 10
  rank -= 3 unless user.logged_in
  rank += 10 if stars == 5 || stars == 1
  if rank > 1 && obscene?(title)
    raise "Underage swearing!" unless user.age > 13
    rank = 1
  elsif rank > 3 && obscene?(text)
    rank = 3 unless user.occupation == "sailor"
  end
  return rank
end

# Test
class Quixotic < SmellTest
  def test_5_star_member
    user = OpenStruct.new(logged_in: true)

    result = rank_hotel_review(user, "title", "body", 5)

    assert_equal 20, result
  end

  def test_3_star_anonymous
    user = OpenStruct.new(logged_in: false)

    result = rank_hotel_review(user, "title", "body", 3)

    assert_equal 7, result
  end
end

# Fake production implementations to simplify example test of subject
def obscene?(text)
  text.include?("obscenities")
end

File: smells/insufficient/quixotic/quixotic.js
// Subject under test
function rankHotelReview (user, title, text, stars) {
  var rank = 10
  if (!user.loggedIn) rank -= 3
  if (stars === 5 || stars === 1) rank += 10
  if (rank > 1 && isObscene(title)) {
    if (user.age <= 13) throw new Error('Underage swearing!')
    rank = 1
  } else if (rank > 3 && isObscene(text)) {
    if (user.occupation !== 'sailor') rank = 3
  }
  return rank
}

// Test
module.exports = {
  fiveStarMember: function () {
    var user = {loggedIn: true}

    var result = rankHotelReview(user, 'title', 'body', 5)

    assert.equal(result, 20)
  },

  threeStarAnonymous: function () {
    var user = {loggedIn: false}

    var result = rankHotelReview(user, 'title', 'body', 3)

    assert.equal(result, 7)
  }
}

// Fake production implementations to simplify example test of subject
function isObscene (text) {
  return text.indexOf('obscenities') !== -1
}
